File: mobile/customer-app/app/src/main/java/com/sukashawarma/customer/ui/screens/outlet/OutletPickerViewModel.kt
package com.sukashawarma.customer.ui.screens.outlet

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.sukashawarma.customer.data.Repository
import com.sukashawarma.customer.data.api.GatewayError
import com.sukashawarma.customer.data.api.GatewayResult
import com.sukashawarma.customer.data.api.OutletDto
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch

data class OutletPickerState(
    val memuat: Boolean = true,
    val galat: GatewayError? = null,
    val kueri: String = "",
    val semua: List<OutletDto> = emptyList(),
    val tampil: List<OutletDto> = emptyList()
)

/**
 * Menyaring outlet berdasarkan nama atau alamat.
 *
 * Fungsi murni, di luar ViewModel, supaya bisa diuji tanpa Android.
 */
fun saringOutlet(outlets: List<OutletDto>, kueri: String): List<OutletDto> {
    val bersih = kueri.trim().lowercase()
    if (bersih.isEmpty()) return outlets
    return outlets.filter {
        it.name.lowercase().contains(bersih) ||
            (it.address?.lowercase()?.contains(bersih) == true)
    }
}

/**
 * Mengurutkan outlet: yang buka di atas, lalu alfabetis.
 *
 * Artboard mengurutkan berdasarkan jarak. Gateway tidak mengirim koordinat
 * pelanggan dan aplikasi belum meminta izin lokasi, jadi jarak tidak dihitung
 * di sini sama sekali -- bukan diisi angka perkiraan.
 */
fun urutkanOutlet(outlets: List<OutletDto>): List<OutletDto> =
    outlets.sortedWith(compareByDescending<OutletDto> { it.isActive }.thenBy { it.name.lowercase() })

class OutletPickerViewModel(private val repository: Repository) : ViewModel() {

    private val _state = MutableStateFlow(OutletPickerState())
    val state: StateFlow<OutletPickerState> = _state.asStateFlow()

    init {
        muat()
    }

    fun muat() {
        _state.value = _state.value.copy(memuat = true, galat = null)
        viewModelScope.launch {
            when (val hasil = repository.outlets()) {
                is GatewayResult.Gagal ->
                    _state.value = _state.value.copy(memuat = false, galat = hasil.error)
                is GatewayResult.Sukses -> {
                    val urut = urutkanOutlet(hasil.data)
                    _state.value = _state.value.copy(
                        memuat = false,
                        galat = null,
                        semua = urut,
                        tampil = saringOutlet(urut, _state.value.kueri)
                    )
                }
            }
        }
    }

    fun ubahKueri(kueri: String) {
        _state.value = _state.value.copy(
            kueri = kueri,
            tampil = saringOutlet(_state.value.semua, kueri)
        )
    }
}

File: mobile/customer-ios-app/SukaShawarma/UI/Home/OutletPickerViewModel.swift
import Foundation
import Observation

/// Padanan `OutletPickerViewModel.kt`.
@MainActor
@Observable
final class OutletPickerViewModel {
    private(set) var memuat = true
    private(set) var galat: GatewayError?
    var kueri = "" { didSet { tampil = saringOutlet(semua, kueri: kueri) } }
    private(set) var semua: [OutletDto] = []
    private(set) var tampil: [OutletDto] = []

    @ObservationIgnored private let repository: Repository

    init(repository: Repository) {
        self.repository = repository
    }

    func muat() async {
        memuat = true
        galat = nil
        switch await repository.outlets() {
        case .gagal(let g):
            memuat = false
            galat = g
        case .sukses(let daftar):
            semua = urutkanOutlet(daftar)
            tampil = saringOutlet(semua, kueri: kueri)
            memuat = false
        }
    }
}

/// Menyaring outlet berdasarkan nama atau alamat.
func saringOutlet(_ outlets: [OutletDto], kueri: String) -> [OutletDto] {
    let bersih = kueri.trimmingCharacters(in: .whitespaces).lowercased()
    guard !bersih.isEmpty else { return outlets }
    return outlets.filter {
        $0.name.lowercased().contains(bersih) || ($0.address?.lowercased().contains(bersih) ?? false)
    }
}

/// Yang buka di atas, lalu alfabetis. Jarak TIDAK dihitung: gateway tak
/// mengirim lokasi pelanggan dan aplikasi belum meminta izin lokasi.
func urutkanOutlet(_ outlets: [OutletDto]) -> [OutletDto] {
    outlets.sorted {
        if $0.isActive != $1.isActive { return $0.isActive }
        return $0.name.lowercased() < $1.name.lowercased()
    }
}
